tutor/src/models/course/period.js

```diff
@@ -51,10 +51,42 @@ class CoursePeriod extends BaseModel {
   unarchive() {
     return { id: this.id, data: { is_archived: false } };
   }
-  afterCreate({ data }) {
+  @action afterCreate({ data }) {
     this.update(data);
     this.course.periods.push(this);
   }
 
+  @action async becomeStudent() {
+    let role = this.course.roles.find((r) => (
+      r.isTeacherStudent && r.period_id == this.id
+    ));
+    if (!role) {
+      const { data } = await this.createTeacherStudent();
+      role = this.course.roles.find(r => r.id == data.id );
+    }
+    await role.become();
+  }
+
+  createTeacherStudent() {
+    return { courseId: this.course.id, id: this.id };
+  }
+
+  @action onCreateTeacherStudent({ data }) {
+    this.course.roles.push(data);
+    this.course.current_role_id = data.id;
```

Comment:
Technically this part doesn't happen until you call `become`...
And `onBecomeSuccess` already does this so I imagine you could remove it?